File: apps/frontend/lib/src/features/iam/domain/entities/role_entity.dart
import 'package:equatable/equatable.dart';

class RoleEntity extends Equatable {
  const RoleEntity({
    required this.id,
    required this.companyId,
    required this.name,
    this.description,
    this.hierarchyLevel = 0,
    this.parentRoleId,
    this.permissionIds = const [],
  });

  final String id;
  final String companyId;
  final String name;
  final String? description;
  final int hierarchyLevel;
  final String? parentRoleId;
  final List<String> permissionIds;

  @override
  List<Object?> get props => [
        id,
        companyId,
        name,
        description,
        hierarchyLevel,
        parentRoleId,
        permissionIds,
      ];
}

File: apps/backend/src/modules/iam/entities/role.entity.ts
import {
  Entity,
  Column,
  Index,
  OneToMany,
  ManyToOne,
  JoinColumn,
} from 'typeorm';
import { TenantBaseEntity } from '../../../shared/database/tenant-base.entity';
import { RolePermission } from './role-permission.entity';
import { UserRole } from './user-role.entity';

@Entity('roles')
@Index(['companyId', 'name'], { unique: true })
export class Role extends TenantBaseEntity {
  @Column({ type: 'varchar', length: 100 })
  name!: string;

  @Column({ type: 'text', nullable: true })
  description?: string;

  @Column({ type: 'int', default: 0, name: 'hierarchy_level' })
  hierarchyLevel!: number;

  @Column({ type: 'uuid', nullable: true, name: 'parent_role_id' })
  parentRoleId?: string;

  @ManyToOne(() => Role, (role) => role.childRoles, { nullable: true })
  @JoinColumn({ name: 'parent_role_id' })
  parentRole?: Role;

  @OneToMany(() => Role, (role) => role.parentRole)
  childRoles!: Role[];

  @OneToMany(() => RolePermission, (rolePermission) => rolePermission.role, {
    cascade: true,
  })
  rolePermissions!: RolePermission[];

  @OneToMany(() => UserRole, (userRole) => userRole.role, { cascade: true })
  userRoles!: UserRole[];
}

